server/src/test/java/org/elasticsearch/action/ActionListenerTests.java

```diff
@@ -370,6 +372,53 @@ public void onFailure(Exception e) {
         assertThat(exReference.get(), instanceOf(IllegalArgumentException.class));
     }
 
+    public void testAssertAtLeastOnceWillLogAssertionErrorWhenNotResolved() throws Exception {
+        assumeTrue("assertAtLeastOnce will be a no-op when assertions are disabled", Assertions.ENABLED);
+        ActionListener<Object> listenerRef = ActionListener.assertAtLeastOnce(ActionListener.running(() -> {
+            // Do nothing, but don't use ActionListener.noop() as it'll never be garbage collected
+        }));
+        // Nullify reference so it becomes unreachable
+        listenerRef = null;
+        assertBusy(() -> {
+            System.gc();
+            assertLeakDetected("LEAK: resource was not cleaned up before it was garbage-collected\\.(.*|\\s)*");
+        });
+    }
+
+    public void testAssertAtLeastOnceWillNotLogWhenResolvedOrFailed() {
+        assumeTrue("assertAtLeastOnce will be a no-op when assertions are disabled", Assertions.ENABLED);
+        ReachabilityChecker reachabilityChecker = new ReachabilityChecker();
+        ActionListener<Object> listenerRef = reachabilityChecker.register(ActionListener.assertAtLeastOnce(ActionListener.running(() -> {
+            // Do nothing, but don't use ActionListener.noop() as it'll never be garbage collected
+        })));
+        // Call onResponse or onFailure
+        if (randomBoolean()) {
+            listenerRef.onResponse("succeeded");
+        } else {
+            listenerRef.onFailure(new RuntimeException("Failed"));
+        }
+        // Nullify reference so it becomes unreachable
+        listenerRef = null;
+        reachabilityChecker.ensureUnreachable();
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testAssertAtLeastOnceWillDelegateResponses() {
+        ActionListener<Object> delegate = mock(ActionListener.class);
+        ActionListener<Object> listener = ActionListener.assertAtLeastOnce(delegate);
+        listener.onResponse("succeeded");
+        verify(delegate).onResponse("succeeded");
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testAssertAtLeastOnceWillDelegateFailures() {
+        ActionListener<Object> delegate = mock(ActionListener.class);
+        ActionListener<Object> listener = ActionListener.assertAtLeastOnce(delegate);
+        RuntimeException exception = new RuntimeException();
+        listener.onFailure(exception);
+        verify(delegate).onFailure(exception);
+    }
```

Comment:
Could we do this without mocks? Mocks tend to get in the way of future refactorings, best avoided IMO unless there really is no other choice. Something like this perhaps?

```suggestion
    public void testAssertAtLeastOnceWillDelegateResponses() {
        final var response = new Object();
        assertSame(
            response,
            safeAwait(SubscribableListener.newForked(l -> ActionListener.assertAtLeastOnce(l).onResponse(response)))
        );
    }

    public void testAssertAtLeastOnceWillDelegateFailures() {
        final var exception = new RuntimeException();
        assertSame(
            exception,
            safeAwaitFailure(SubscribableListener.newForked(l -> ActionListener.assertAtLeastOnce(l).onFailure(exception)))
        );
    }
```